// @cashscript/utils/dist/data.js
import { bigIntToVmNumber, utf8ToBin, binToUtf8, vmNumberToBigInt, isVmNumberError, } from '@bitauth/libauth';
export function encodeBool(bool) {
    return bool ? encodeInt(1n) : encodeInt(0n);
}
export function decodeBool(encodedBool) {
    // Any encoding of 0 is false, else true
    for (let i = 0; i < encodedBool.byteLength; i += 1) {
        if (encodedBool[i] !== 0) {
            // Can be negative zero
            if (i === encodedBool.byteLength - 1 && encodedBool[i] === 0x80)
                return false;
            return true;
        }
    }
    return false;
}
export function encodeInt(int) {
    return bigIntToVmNumber(int);
}
export function decodeInt(encodedInt, maxLength = 8) {
    const options = { maximumVmNumberByteLength: maxLength };
    const result = vmNumberToBigInt(encodedInt, options);
    if (isVmNumberError(result)) {
        throw new Error(result);
    }
    return result;
}
export function encodeString(str) {
    return utf8ToBin(str);
}
export function decodeString(encodedString) {
    return binToUtf8(encodedString);
}
export function placeholder(size) {
    return new Uint8Array(size).fill(0);
}
